plugins/genesys/fu-genesys-usbhub-device.c

```diff
@@ -605,6 +613,61 @@ fu_genesys_usbhub_device_get_fw_size(FuGenesysUsbhubDevice *self, int bank_num,
 	/* success */
 	return TRUE;
 }
+
+static gboolean
+fu_genesys_usbhub_device_get_fw_version(FuDevice *device, int bank_num, GError **error)
+{
+	FuGenesysUsbhubDevice *self = FU_GENESYS_USBHUB_DEVICE(device);
+	gsize address = self->fw_bank_addr[bank_num];
+	gsize bufsz;
+	guint16 version_raw;
+	g_autoptr(FuFirmware) firmware = NULL;
+	g_autofree guint8 *buf = NULL;
+
+	g_return_val_if_fail(bank_num < 2, FALSE);
+
+	if (self->code_size[bank_num] == 0) {
+		if (!fu_genesys_usbhub_device_get_fw_size(self, bank_num, error))
+			return FALSE;
+	} else {
+		if (!fu_genesys_usbhub_device_check_fw_signature(self, bank_num, error))
+			return FALSE;
+	}
+
+	bufsz = self->code_size[bank_num];
+	buf = g_malloc0(bufsz);
+	if (!fu_genesys_usbhub_device_read_flash(self, address, buf, bufsz, NULL, error))
+		return FALSE;
+
+	firmware = fu_genesys_usbhub_firmware_new();
+	if (!fu_firmware_parse(firmware,
+			       g_bytes_new_take(buf, bufsz),
+			       FWUPD_INSTALL_FLAG_NONE,
+			       error)) {
+		self->fw_bank_vers[bank_num] = 0;
+		return FALSE;
+	}
+
+	version_raw = fu_firmware_get_version_raw(firmware);
+	if (version_raw != 0xffff)
+		self->fw_bank_vers[bank_num] = version_raw;
+
+	return TRUE;
+}
+
+static guint8
+_fu_genesys_revert_hex_char_to_int(guint8 var)
```

Comment:
I think you are re-implementing `g_ascii_xdigit_value()`.

If so, I prefer you reuse it.

Reply:
`_fu_genesys_revert_hex_char_to_int()` converts extend char 'g'~'z', because the string descriptor replay as UTF-16 string format as default. But we use the tool string to contain some info.

Reply:
Oh I missed to till 'Z' range.

So I guess we can reuse the glib code and hack it a bit for the z range.

I would suggest the following:

```
static gint
fu_genesys_tsdigit_value (gchar c)
{
	if (c >= 'A' && c <= 'Z')
		return c - 'A' + 10;
	if (c >= 'a' && c <= 'f')
		return c - 'a' + 10;
	return g_ascii_digit_value (c);                                               
}
```

This way, the function returns -1 if the character is not a toolstring digit; and to some paranoïd testing accross the _setup() vfunc.

What do you think?